src/structure.ts

```diff
@@ -81,44 +81,46 @@ const isCommand = (value: any, warn = false): value is Command => {
     return true;
 };
 
+//type AccType = { [key: string]: Command };
+
 /**
  * Validate that a set of values are {@link Command} objects
  */
-export const validateCommands = <Ctx extends Context = Context, Req extends Request = Request, Sentry extends Toucan | undefined = undefined>(cmds: any[], warn = false) =>
-    cmds.reduce((acc, cmd) => {
-        if (!isCommand(cmd, warn)) return acc;
+export const validateCommands = <Ctx extends Context = Context, Req extends Request = Request, Sentry extends Toucan | undefined = undefined>(cmds: unknown[], warn = false) =>
+cmds.reduce((acc: Record<string, Command>, cmd: unknown) => {
```

Comment:
```suggestion
cmds.reduce<Commands<Ctx, Req, Sentry>>((acc, cmd) => {
```
A couple of related thoughts here:

1. We have the `Commands` type that already represents the Record, and we should pass in the generics
2. I believe reduce can take a generic type like this, so you don't need to type the params of the method -- I believe the `as` at the end can also be removed?

Reply:
I tried playing around with this a little bit and it threw an error when it was handled this way:

```ts
export const validateCommands = <Ctx extends Context = Context, Req extends Request = Request, Sentry extends Toucan | undefined = undefined>(cmds: unknown[], warn = false) =>
cmds.reduce<Commands<Ctx, Req, Sentry>>((acc, cmd) => {
    if (!isCommand(cmd, warn)) return acc;

    // Check the command doesn't already exist
    if (acc[cmd.name]) {
        if (warn)
            console.warn(`Command ${cmd.name} already exists`);
        return acc;
    }

    // Add the command
    return {
        ...acc,
        [cmd.name]: cmd,
    };
}, {});
```

This is the error, it's thrown at `(acc, cmd) => {` for both `validateCommands` and `validateComponents`.

```
Argument of type '(acc: Commands<Ctx, Req, Sentry>, cmd: unknown) => { [x: string]: Command<Context, Request<unknown, CfProperties<unknown>>, undefined> | Command<...>; }' is not assignable to parameter of type '(previousValue: Commands<Ctx, Req, Sentry>, currentValue: unknown, currentIndex: number, array: unknown[]) => Commands<Ctx, Req, Sentry>'.
  Type '{ [x: string]: Command<Context, Request<unknown, CfProperties<unknown>>, undefined> | Command<Ctx, Req, Sentry>; }' is not assignable to type 'Commands<Ctx, Req, Sentry>'.
    'string' index signatures are incompatible.
      Type 'Command<Context, Request<unknown, CfProperties<unknown>>, undefined> | Command<Ctx, Req, Sentry>' is not assignable to type 'Command<Ctx, Req, Sentry>'.
        Type 'Command<Context, Request<unknown, CfProperties<unknown>>, undefined>' is not assignable to type 'Command<Ctx, Req, Sentry>'.
          Type 'Sentry' is not assignable to type 'undefined'.
            Type 'Toucan | undefined' is not assignable to type 'undefined'.
              Type 'Toucan' is not assignable to type 'undefined'.ts(2345)
```

I tried to play off this a little bit by doing this:

```ts
export const validateCommands = <Ctx extends Context = Context, Req extends Request = Request, Sentry extends Toucan | undefined = undefined>(cmds: unknown[], warn = false) =>
cmds.reduce<Record<string, Command>>((acc, cmd) => {
    if (!isCommand(cmd, warn)) return acc;

    // Check the command doesn't already exist
    if (acc[cmd.name]) {
        if (warn)
            console.warn(`Command ${cmd.name} already exists`);
        return acc;
    }

    // Add the command
    return {
        ...acc,
        [cmd.name]: cmd,
    };
}, {}) as Commands<Ctx, Req, Sentry>;
```

Which doesn't throw any errors, though it's still similar to the implementation I tried previously. Would you want me to just add it this way or try some additional changes?

Reply:
Ahh, `isCommand` might need to be updated to handle the same generics?

Reply:
Yeah seems like it, I'll see what I can do.

Reply:
Changed the headers to this:

```ts
function isComponent<Ctx extends Context, Req extends Request, Sentry extends Toucan | undefined>(value: unknown, warn: boolean): value is Component<Ctx, Req, Sentry> {
```

which fixed the errors, I'll push my changes now.